### PR TITLE
RATIS-2045. SnapshotInstallationHandler doesn't notify follower when snapshotIndex is -1 and firstAvailableLogIndex is 0

### PR DESCRIPTION
**Pull request by hemantk-12**
## What changes were proposed in this pull request?
When the leader has firstAvailableLogIndex `0` and the follower node is bootstrapping and doesn't have any snapshot, the leader doesn't notify the follower to install the snapshot because [the current condition](https://github.com/apache/ratis/blob/44a4f1fb73a073589a4563d492f6acfc68ce4a5d/ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java#L224) is always `true` if `snapshotIndex` is `-1` (default value) and `firstAvailableLogIndex` is `0` (when the leader has logs form 0th index).
As part of this change, we are changing the condition to notify when `snapshotIndex` is `-1`. If `snapshotIndex` is not `-1`, then continue with `snapshotIndex + 1 >= firstAvailableLogIndex && firstAvailableLogIndex > INVALID_LOG_INDEX`.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/RATIS-2045

## How was this patch tested?
Existing unit tests for now.

### CODE REVIEW

**Comment by szetszwo** [comment]
@duongkame , there are some test failures related to `IllegalState allLeaks.size = 2`; see https://github.com/apache/ratis/actions/runs/8350541481/job/22857177939?pr=1053#step:6:1096

**Comment by duongkame** [comment]
> @duongkame , there are some test failures related to `IllegalState allLeaks.size = 2`; see https://github.com/apache/ratis/actions/runs/8350541481/job/22857177939?pr=1053#step:6:1096

I created a custom build with advance detection here: https://github.com/duongkame/ratis/actions/runs/8351812136

Looks like this is because the [retained log entries](https://github.com/apache/ratis/blob/master/ratis-server/src/test/java/org/apache/ratis/statemachine/impl/SimpleStateMachine4Testing.java#L203-L203) in SimpleStateMachine4Testing cache are not released. Probably because the state machine is not closed. Filed RATIS-2046.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java [hunk 1]
```diff
@@ -353,4 +354,4 @@ private RoleInfoProto getRoleInfoProto(RaftPeer leader) {
         .setFollowerInfo(followerInfo)
         .build();
   }
-}
\ No newline at end of file
+}
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java:357]:
```suggestion
}

```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java [hunk 2]
```diff
@@ -353,4 +354,5 @@ private RoleInfoProto getRoleInfoProto(RaftPeer leader) {
...
+
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java:358]:
```suggestion
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java (hunk 1 above)
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java:357]: @hemantk-12 , could you revert this whitespace change?  I tried to do in on the GitHub web ui but it did not work.

Comment (hemantk-12) [reply ratis-server/src/main/java/org/apache/ratis/server/impl/SnapshotInstallationHandler.java]: Removed.

**Review by szetszwo — APPROVED** [review]
+1

**Pull request merged**